### PR TITLE
Simplifying atom checking and add %

### PR DESCRIPTION
**Pull request by rgrinberg**
@diml did you have a reason to keep these 2 slightly different definitions?

Btw, we could implement `Atom.is_valid` it the lexer, but it will cost us a bit of performance. Having some sort of trivial meta programming like cinaps might solve this issue.

### CODE REVIEW

**Comment by rgrinberg** [comment]
So this change actually breaks Camomile as it uses `%` in symbols. We definitely need better error handling here, but moreover, how do we preserve compat here. We can't really make the definition of an atom dependent on whether we are in a jbuild or a dune file.

One thing that we can do which will be consistent for dune and jbuild files would be to allow `%` in symbol names as long as it isn't followed by a `{`. This will be consistent with how we'd handle `$` in jbuild files anyway.

**Comment by diml** [comment]
Allowing `%` unescaped when it is not followed by `{` seems good. For compatibility, these functions are only used for generating s-expressions so it seems fine if they can only generate the latest supported version.

BTW, the two functions here where not exactly the same. The first one allows control characters but not the second one. This is because the second onenis supposed to produce a readable s-expression, while the first is just for checking validity.

**Comment by rgrinberg** [comment]
I see. It still seems like a good idea to exclude non ascii chars from being part of atoms. Atoms are a bit like lisp symbols, which should be readable.

**Comment by diml** [comment]
That seems reasonable to me. We should change the lexer as well then.

**Comment by rgrinberg** [comment]
@diml I think I did in this PR. I've excluded all `'127' - '255'` chars from the `atom_char` set. I still need to fix the escaping for `%` though.

**Comment by diml** [comment]
Ah sorry, I missed that. So this PR just needs to be updated to allow `%` unescaped if it's not followed by `{`, right?

**Comment by rgrinberg** [comment]
Yeah, exactly. Should we allow `%{` in jbuild files btw?

**Comment by diml** [comment]
Yes, actually the syntax in jbuild files should be unchanged. I missed that this was modifying the old syntax as well. This makes me more confident that we should split the lexers for the old and new syntax.

Now I see that there is indeed an inconsistency between what the lexer produces and the atom invariant. I can think of two solutions:
1. we map the output of the old lexer to transparently turn old atoms that are now invalid into quoted strings. The worst that can happen is an invalid error message
2. we change the model a bit: so that in the memory representation we have no restrictions on the contents of atoms and quoted strings and the printer can decide to turn atoms into quoted strings if needed

If we assume that we are not going to do major changes to the lexical conventions in the future, (1) seems simpler to me.

File: src/usexp/lexer.mll [hunk 1]
```diff
@@ -75,10 +75,14 @@ type block_string_line_kind =
...
 let newline   = '\r'? '\n'
 let blank     = [' ' '\t' '\012']
-let atom_char = [^ ';' '(' ')' '"' ' ' '\t' '\r' '\n' '\012']
 let digit     = ['0'-'9']
 let hexdigit  = ['0'-'9' 'a'-'f' 'A'-'F']
 
+let atom_char_jbuild =
+  [^ ';' '(' ')' '"' ' ' '\t' '\r' '\n' '\012']
+let atom_char_dune =
+  [^ ';' '(' ')' '"' ' ' '\t' '\r' '\n' '\000'-'\032' '\127'-'\255']
```
Comment (diml) [inline src/usexp/lexer.mll]: this can be slightly simplified as `' '` and `\{t,r,n}` are in  `\000-\032`

**Comment by rgrinberg** [comment]
@diml I simplified the char set. Will merge modulo the last commit.

**Pull request merged**